src/DynamoCore/Core/Threading/UpdateGraphAsyncTask.cs

```diff
@@ -103,37 +120,37 @@ protected override void HandleTaskCompletionCore()
             }
         }
 
+        /// <summary>
+        /// Return if this task's graph sync data is a super set of the other
+        /// </summary>
+        /// <param name="other"></param>
+        /// <returns></returns>
+        private bool Contains(UpdateGraphAsyncTask other)
+        {
+            // Be more conservative here. Not only check ModifiedNodes, but
+            // also check *all* nodes in graph sync data. For example, consider
+            // a CBN outputs to a node, the node is a downstream node of cbn.
+            //
+            // Now if node is modified and request a run, task's ModifiedNodes
+            // will cotain this node; then CBN is modified and request a run,
+            // ModifiedNodes would contain both CBN and the node, and previous
+            // task will be thrown away.
+            return other.graphSyncData.NodeIDs.All(graphSyncData.NodeIDs.Contains) && 
+                   other.ModifiedNodes.All(ModifiedNodes.Contains); 
+        }
```

Comment:
Does this not remain the same for the following use case in your example? It seems like the second `UpdateGraphAsyncTask` will still throw the first one away since the second task is a superset of the first?

      cbn                 foo          task queue
    +------+           +--------+      UpdateGraphAsyncTask: {ModifiedNodes: {foo}, Code: {foo(x)}}
    |  42  | --------> | foo(x) |      UpdateGraphAsyncTask: {ModifiedNodes: {cbn, foo}, Code: {t = 42}}
    +------+           +--------+